plugins/node/opentelemetry-instrumentation-fastify/src/instrumentation.ts

```diff
@@ -0,0 +1,248 @@
+      if (!instrumentation.isEnabled()) {
+        return original.apply(this, args);
+      }
+
+      const spanName = `${FastifyNames.MIDDLEWARE} - ${
+        original.name || ANONYMOUS_NAME
+      }`;
+
+      const span = instrumentation._startSpan(spanName, {
+        [AttributeNames.FASTIFY_TYPE]: FastifyTypes.MIDDLEWARE,
+        [AttributeNames.PLUGIN_NAME]: pluginName,
+        [AttributeNames.HOOK_NAME]: hookName,
+      });
+
+      return safeExecuteInTheMiddleMaybePromise(
+        () => {
+          return original.apply(this, args);
+        },
+        err => {
+          if (err) {
+            span.setStatus({
+              code: SpanStatusCode.ERROR,
+              message: err.message,
+            });
+            span.recordException(err);
+          }
+          span.end();
+        }
+      );
+    };
+  }
+
+  private _wrapAddHook(): (
+    original: FastifyInstance['addHook']
+  ) => () => FastifyInstance {
+    const instrumentation = this;
+    return function (
+      original: FastifyInstance['addHook']
+    ): () => FastifyInstance {
+      return function wrappedAddHook(this: any, ...args: any) {
+        if (!instrumentation.isEnabled()) {
+          return original.apply(this, args);
+        }
+
+        const name = args[0] as string;
+        const handler = args[1] as HandlerOriginal;
+        const pluginName = this.pluginName;
+        if (name === 'onRegister') {
+          return original.apply(this, [name as any, handler]);
+        }
+        return original.apply(this, [
+          name as any,
+          instrumentation._wrapHandler(pluginName, name, handler),
+        ]);
+      };
+    };
+  }
+
+  private _patchConstructor(
+    original: () => FastifyInstance
+  ): () => FastifyInstance {
+    const instrumentation = this;
+    return function (this: FastifyInstance, ...args) {
+      const app: FastifyInstance = original.apply(this, args);
+      app.addHook('onRequest', instrumentation._hookOnRequest());
+      app.addHook('preHandler', instrumentation._hookPreHandler());
+
+      instrumentation._wrap(app, 'addHook', instrumentation._wrapAddHook());
+
+      return app;
+    };
+  }
+
+  public _patchSend(span: Span) {
+    const instrumentation = this;
+    return function patchSend(
+      original: () => FastifyReply
+    ): () => FastifyReply {
+      return function send(this: FastifyReply, ...args) {
+        if (!instrumentation.isEnabled()) {
+          return original.apply(this, args);
+        }
+
+        return safeExecuteInTheMiddle<FastifyReply>(
+          () => {
+            return original.apply(this, args);
+          },
+          err => {
+            if (err) {
+              span.setStatus({
+                code: SpanStatusCode.ERROR,
+                message: err.message,
+              });
+              span.recordException(err);
+            }
+            span.end();
+          }
+        );
+      };
+    };
+  }
+
+  public _hookPreHandler() {
```

Comment:
Why are you wrapping only this hook specifically?

As a user, I do not think I would like any of my hooks to create spans. In case I want the hooks to be wrapped in spans.. then I want all of them.

Reply:
I'm not sure if I understand you. First this is not wrapping, but just adding a preHandler hook to be able to catch the beginning of user defined route, all hooks are wrapped in different place